TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Model/Workspace.php

```diff
@@ -547,95 +570,75 @@ protected function moveNodeVariantToTargetWorkspace(NodeInterface $node, Workspa
+        if ($nodeInTargetWorkspaceBase !== null && $nodeInTargetWorkspaceBase->getPath() === $sourceNodeData->getPath()) {
+            $this->nodeDataRepository->remove($sourceShadowNodeData);
             return;
         }
 
-        $shadowNodeData->setMovedTo($targetNodeData);
-        $shadowNodeData->setWorkspace($targetWorkspace);
+        $sourceShadowNodeData->setMovedTo($targetNodeData);
+        $sourceShadowNodeData->setWorkspace($targetWorkspace);
 
-        if ($nodeInTargetWorkspaceBase !== null && $nodeInTargetWorkspaceBase->getPath() !== $shadowNodeData->getPath()) {
-            $this->adjustShadowNodePath($shadowNodeData, $nodeInTargetWorkspaceBase->getPath(), $targetWorkspace, $publishedNodeData->getDimensionValues());
+        if ($nodeInTargetWorkspaceBase !== null && $nodeInTargetWorkspaceBase->getPath() !== $sourceShadowNodeData->getPath()) {
+            $nodeOnSamePathInTargetWorkspace = $this->nodeDataRepository->findOneByPath($nodeInTargetWorkspaceBase->getPath(), $targetWorkspace, $sourceNodeData->getDimensionValues());
+            if ($nodeOnSamePathInTargetWorkspace === null || $nodeOnSamePathInTargetWorkspace->getWorkspace() !== $targetWorkspace) {
+                $sourceShadowNodeData->setPath($nodeInTargetWorkspaceBase->getPath(), false);
+            } else {
+                // A node exists in that path, so no shadow node is needed/possible.
+                $this->nodeDataRepository->remove($sourceShadowNodeData);
+            }
         }
-    }
 
-    /**
-     * Adjusts the path of $shadowNodeData to $path, if needed/possible.
-     *
-     * If the $path is occupied in $targetWorkspace, the shadow is removed.
-     *
-     * @param NodeData $shadowNodeData
-     * @param $path
-     * @param Workspace $targetWorkspace
-     * @param array $dimensionValues
-     * @return void
-     */
-    protected function adjustShadowNodePath(NodeData $shadowNodeData, $path, Workspace $targetWorkspace, array $dimensionValues)
-    {
-        $nodeOnSamePathInTargetWorkspace = $this->nodeDataRepository->findOneByPath($path, $targetWorkspace, $dimensionValues);
-        if ($nodeOnSamePathInTargetWorkspace === null || $nodeOnSamePathInTargetWorkspace->getWorkspace() !== $targetWorkspace) {
-            $shadowNodeData->setPath($path, false);
-            return;
+        // Check if a shadow node which has the same path, workspace and dimension values like the shadow node data we just created already exists (in the target workspace).
+        // If it does, we re-use the existing node and make sure that all properties etc. are taken from the node which is being published.
+        $existingShadowNodeDataInTargetWorkspace = $this->nodeDataRepository->findOneByPath($sourceShadowNodeData->getPath(), $sourceShadowNodeData->getWorkspace(), $sourceShadowNodeData->getDimensionValues(), true);
```

Comment:
Yes, it should be `$targetWorkspace` to be correct. It does work though already, because $sourceShadowNodeData->getWorkspace()` is the target workspace at this point, because `$sourceShadowNodeData` was already moved to the new workspace in line 621.

>  Shouldn't the resulting (shadow) nodedata also be checked on being in exactly the given workspace and not one of the base workspaces of those?

Yes, we should check that. I think that `findOneByPath` with workspace as one of the arguments is dangerously misleading. It does not tell you that you might get a node returned which is from a different workspace.